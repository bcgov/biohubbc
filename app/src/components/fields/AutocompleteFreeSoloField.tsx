import TextField from '@material-ui/core/TextField';
import Autocomplete, { createFilterOptions } from '@material-ui/lab/Autocomplete';
import { useFormikContext } from 'formik';
import React from 'react';

export interface IAutocompleteFreeSoloField {
  id: string;
  label: string;
  name: string;
  options: string[];
  required?: boolean;
  filterLimit?: number;
}

const AutocompleteFreeSoloField: React.FC<IAutocompleteFreeSoloField> = (props) => {
  const { touched, errors, setFieldValue, values } = useFormikContext<string>();

  return (
    <Autocomplete
      freeSolo
      autoSelect
      includeInputInList
      clearOnBlur
      blurOnSelect
      handleHomeEndKeys
      id={props.id}
      data-testid={props.id}
      value={values[props.name]}
      options={props.options}
      getOptionLabel={(option) => option}
      filterOptions={createFilterOptions({ limit: props.filterLimit })}
      onChange={(event, option) => {
        setFieldValue(props.name, option);
      }}
      renderInput={(params) => (
        <TextField
          {...params}
          required={props.required}
          label={props.label}
          variant="outlined"
          fullWidth
          error={touched[props.name] && Boolean(errors[props.name])}
          helperText={touched[props.name] && errors[props.name]}
        />
      )}
    />
  );
};

export default AutocompleteFreeSoloField;
