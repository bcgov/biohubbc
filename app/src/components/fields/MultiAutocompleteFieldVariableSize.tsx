import Checkbox from '@material-ui/core/Checkbox';
import ListSubheader from '@material-ui/core/ListSubheader';
import TextField from '@material-ui/core/TextField';
import CheckBox from '@material-ui/icons/CheckBox';
import CheckBoxOutlineBlank from '@material-ui/icons/CheckBoxOutlineBlank';
import Autocomplete from '@material-ui/lab/Autocomplete';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { useFormikContext } from 'formik';
import React from 'react';
import { ListChildComponentProps, VariableSizeList } from 'react-window';
import get from 'lodash-es/get';

const LISTBOX_PADDING = 8; // px

export interface IMultiAutocompleteFieldOption {
  value: string | number;
  label: string;
}

export interface IMultiAutocompleteField {
  id: string;
  label: string;
  options: IMultiAutocompleteFieldOption[];
  required?: boolean;
  filterLimit?: number;
}

function renderRow(props: ListChildComponentProps) {
  const { data, index, style } = props;
  return React.cloneElement(data[index], {
    style: {
      ...style,
      top: (style.top as number) + LISTBOX_PADDING
    }
  });
}

const OuterElementContext = React.createContext({});

const OuterElementType = React.forwardRef<HTMLDivElement>((props, ref) => {
  const outerProps = React.useContext(OuterElementContext);
  return <div ref={ref} {...props} {...outerProps} />;
});

function useResetCache(data: any) {
  const ref = React.useRef<VariableSizeList>(null);
  React.useEffect(() => {
    if (ref.current != null) {
      ref.current.resetAfterIndex(0, true);
    }
  }, [data]);
  return ref;
}

// Adapter for react-window
const ListboxComponent = React.forwardRef<HTMLDivElement>(function ListboxComponent(props, ref) {
  const { children, ...other } = props;
  const itemData = React.Children.toArray(children);
  const itemCount = itemData.length;
  const itemSize = 54;

  const getChildSize = (child: React.ReactNode) => {
    if (React.isValidElement(child) && child.type === ListSubheader) {
      return 48;
    }

    return itemSize;
  };

  const getHeight = () => {
    if (itemCount > 8) {
      return 8 * itemSize;
    }
    return itemData.map(getChildSize).reduce((a, b) => a + b, 0);
  };

  const gridRef = useResetCache(itemCount);

  return (
    <div ref={ref}>
      <OuterElementContext.Provider value={other}>
        <VariableSizeList
          itemData={itemData}
          height={getHeight() + 2 * LISTBOX_PADDING}
          width="100%"
          ref={gridRef}
          outerElementType={OuterElementType}
          innerElementType="ul"
          itemSize={(index: number) => getChildSize(itemData[index])}
          overscanCount={5}
          itemCount={itemCount}>
          {renderRow}
        </VariableSizeList>
      </OuterElementContext.Provider>
    </div>
  );
});

const useStyles = makeStyles({
  listbox: {
    boxSizing: 'border-box',
    '& ul': {
      padding: 0,
      margin: 0
    }
  }
});

const MultiAutocompleteFieldVariableSize: React.FC<IMultiAutocompleteField> = (props) => {
  const classes = useStyles();

  const { values, touched, errors, setFieldValue } = useFormikContext<IMultiAutocompleteFieldOption>();

  const getExistingValue = (existingValues: any[]): IMultiAutocompleteFieldOption[] => {
    if (!existingValues) {
      return [];
    }

    return props.options.filter((option) => existingValues.includes(option.value));
  };

  const handleGetOptionSelected = (
    option: IMultiAutocompleteFieldOption,
    value: IMultiAutocompleteFieldOption
  ): boolean => {
    if (!option?.value || !value?.value) {
      return false;
    }

    return option.value === value.value;
  };

  return (
    <Autocomplete
      multiple
      autoHighlight={true}
      value={getExistingValue(get(values, props.id))}
      ListboxComponent={ListboxComponent as React.ComponentType<React.HTMLAttributes<HTMLElement>>}
      id={props.id}
      data-testid={props.id}
      options={props.options}
      getOptionLabel={(option) => option.label}
      getOptionSelected={handleGetOptionSelected}
      disableCloseOnSelect
      disableListWrap
      classes={classes}
      onChange={(event, option) => {
        setFieldValue(
          props.id,
          option.map((item) => item.value)
        );
      }}
      renderOption={(option, { selected }) => {
        const disabled: any = props.options && props.options?.indexOf(option) !== -1;
        return (
          <>
            <Checkbox
              icon={<CheckBoxOutlineBlank fontSize="small" />}
              checkedIcon={<CheckBox fontSize="small" />}
              style={{ marginRight: 8 }}
              checked={selected}
              disabled={disabled}
              value={option.value}
            />
            {option.label}
          </>
        );
      }}
      renderInput={(params) => (
        <TextField
          {...params}
          required={props.required}
          label={props.label}
          variant="outlined"
          fullWidth
          error={get(touched, props.id) && Boolean(get(errors, props.id))}
          helperText={get(touched, props.id) && get(errors, props.id)}
        />
      )}
    />
  );
};

export default MultiAutocompleteFieldVariableSize;
