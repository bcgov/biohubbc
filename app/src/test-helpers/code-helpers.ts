import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';

export const codes: IGetAllCodeSetsResponse = {
  coordinator_agency: [{ id: 1, name: 'A Rocha Canada' }],
  management_action_type: [{ id: 1, name: 'Management action' }],
  first_nations: [{ id: 1, name: 'First nations code' }],
  funding_source: [{ id: 1, name: 'Funding source code' }],
  investment_action_category: [{ id: 1, fs_id: 1, name: 'Investment action category' }],
  activity: [{ id: 1, name: 'Activity code' }],
  project_type: [{ id: 1, name: 'Project type' }],
  region: [{ id: 1, name: 'Region code' }],
  species: [{ id: 1, name: 'Species code' }],
  proprietor_type: [
    { id: 1, name: 'Proprietor code 1', is_first_nation: false },
    { id: 2, name: 'First Nations Land', is_first_nation: true }
  ],
  iucn_conservation_action_level_1_classification: [{ id: 1, name: 'IUCN class' }],
  iucn_conservation_action_level_2_subclassification: [{ id: 1, iucn1_id: 1, name: 'IUCN subclass 1' }],
  iucn_conservation_action_level_3_subclassification: [{ id: 1, iucn2_id: 1, name: 'IUCN subclass 2' }],
  system_roles: [
    { id: 1, name: 'Role 1' },
    { id: 2, name: 'Role 2' }
  ],
  project_roles: [
    { id: 1, name: 'Project Role 1' },
    { id: 2, name: 'Project Role 2' }
  ],
  administrative_activity_status_type: [
    { id: 1, name: 'Pending' },
    { id: 2, name: 'Actioned' },
    { id: 3, name: 'Rejected' }
  ],
  common_survey_methodologies: [
    { id: 1, name: 'Recruitment' },
    { id: 2, name: 'SRB' }
  ]
};
