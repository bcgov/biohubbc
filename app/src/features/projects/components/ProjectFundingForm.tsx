import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Divider from '@material-ui/core/Divider';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Paper from '@material-ui/core/Paper';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { mdiPencilOutline, mdiPlus, mdiTrashCanOutline } from '@mdi/js';
import Icon from '@mdi/react';
import EditDialog from 'components/dialog/EditDialog';
import { IMultiAutocompleteFieldOption } from 'components/fields/MultiAutocompleteFieldVariableSize';
import { DATE_FORMAT } from 'constants/dateTimeFormats';
import { AddFundingI18N } from 'constants/i18n';
import { FieldArray, useFormikContext } from 'formik';
import React, { useState } from 'react';
import { getFormattedAmount, getFormattedDateRangeString } from 'utils/Utils';
import yup from 'utils/YupSchema';
import ProjectFundingItemForm, {
  IProjectFundingFormArrayItem,
  ProjectFundingFormArrayItemInitialValues,
  ProjectFundingFormArrayItemYupSchema
} from './ProjectFundingItemForm';

export interface IProjectFundingForm {
  funding_sources: IProjectFundingFormArrayItem[];
}

export const ProjectFundingFormInitialValues: IProjectFundingForm = {
  funding_sources: []
};

export const ProjectFundingFormYupSchema = yup.object().shape({});

export interface IInvestmentActionCategoryOption extends IMultiAutocompleteFieldOption {
  fs_id: number;
}

export interface IProjectFundingFormProps {
  funding_sources: IMultiAutocompleteFieldOption[];
  investment_action_category: IInvestmentActionCategoryOption[];
}

const useStyles = makeStyles((theme: Theme) => ({
  title: {
    flexGrow: 1,
    marginRight: '1rem',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    fontWeight: 700
  },
  titleDesc: {
    marginLeft: theme.spacing(1),
    fontWeight: 400
  },
  fundingListIem: {
    padding: 0,
    '& + li': {
      marginTop: theme.spacing(2)
    }
  },
  fundingListItemInner: {
    flexGrow: 1,
    flexShrink: 1,
    overflow: 'hidden'
  },
  fundingListItemToolbar: {
    paddingRight: theme.spacing(2)
  }
}));

/**
 * Create project - Funding section
 *
 * @return {*}
 */
const ProjectFundingForm: React.FC<IProjectFundingFormProps> = (props) => {
  const classes = useStyles();

  const formikProps = useFormikContext<IProjectFundingForm>();
  const { values } = formikProps;

  //Tracks information about the current funding source item that is being added/edited
  const [currentProjectFundingFormArrayItem, setCurrentProjectFundingFormArrayItem] = useState({
    index: 0,
    values: ProjectFundingFormArrayItemInitialValues
  });

  const [isModalOpen, setIsModalOpen] = useState(false);

  return (
    <form onSubmit={formikProps.handleSubmit}>
      <Box>
        <Box component="header" display="flex" justifyContent="space-between" alignItems="center" mb={2}>
          <Typography variant="h3">Funding Sources ({values.funding_sources.length})</Typography>
          <Button
            data-testid="add-button"
            variant="outlined"
            color="primary"
            title="Add Funding Source"
            aria-label="Add Funding Source"
            startIcon={<Icon path={mdiPlus} size={1} />}
            onClick={() => {
              setCurrentProjectFundingFormArrayItem({
                index: values.funding_sources.length,
                values: ProjectFundingFormArrayItemInitialValues
              });
              setIsModalOpen(true);
            }}>
            Add Funding Source
          </Button>
        </Box>
        <Box>
          <FieldArray
            name="funding_sources"
            render={(arrayHelpers) => (
              <Box mb={2}>
                <EditDialog
                  dialogTitle={AddFundingI18N.addTitle}
                  open={isModalOpen}
                  component={{
                    element: (
                      <ProjectFundingItemForm
                        funding_sources={props.funding_sources}
                        investment_action_category={props.investment_action_category}
                      />
                    ),
                    initialValues: currentProjectFundingFormArrayItem.values,
                    validationSchema: ProjectFundingFormArrayItemYupSchema
                  }}
                  onCancel={() => setIsModalOpen(false)}
                  onSave={(projectFundingItemValues) => {
                    if (currentProjectFundingFormArrayItem.index < values.funding_sources.length) {
                      // Update an existing item
                      arrayHelpers.replace(currentProjectFundingFormArrayItem.index, projectFundingItemValues);
                    } else {
                      // Add a new item
                      arrayHelpers.push(projectFundingItemValues);
                    }

                    // Close the modal
                    setIsModalOpen(false);
                  }}
                />
                <List dense disablePadding>
                  {!values.funding_sources.length && (
                    <ListItem dense component={Paper}>
                      <Box display="flex" flexGrow={1} justifyContent="center" alignContent="middle" p={2}>
                        <Typography variant="subtitle2">No Funding Sources</Typography>
                      </Box>
                    </ListItem>
                  )}
                  {values.funding_sources.map((fundingSource, index) => {
                    const investment_action_category_label =
                      (fundingSource.agency_id === 1 && 'Investment Action') ||
                      (fundingSource.agency_id === 2 && 'Investment Category') ||
                      null;

                    const investment_action_category_value = props.investment_action_category.filter(
                      (item) => item.value === fundingSource.investment_action_category
                    )?.[0]?.label;

                    return (
                      <ListItem dense className={classes.fundingListIem} key={index}>
                        <Paper className={classes.fundingListItemInner}>
                          <Toolbar className={classes.fundingListItemToolbar}>
                            <Typography className={classes.title}>
                              {getCodeValueNameByID(props.funding_sources, fundingSource.agency_id)}
                              {investment_action_category_label && (
                                <span className={classes.titleDesc}>({investment_action_category_value})</span>
                              )}
                            </Typography>

                            <IconButton
                              color="primary"
                              data-testid={'edit-button-' + index}
                              title="Edit Funding Source"
                              aria-label="Edit Funding Source"
                              onClick={() => {
                                setCurrentProjectFundingFormArrayItem({
                                  index: index,
                                  values: values.funding_sources[index]
                                });
                                setIsModalOpen(true);
                              }}>
                              <Icon path={mdiPencilOutline} size={1} />
                            </IconButton>
                            <IconButton
                              color="primary"
                              data-testid={'delete-button-' + index}
                              title="Remove Funding Source"
                              aria-label="Remove Funding Source"
                              onClick={() => arrayHelpers.remove(index)}>
                              <Icon path={mdiTrashCanOutline} size={1} />
                            </IconButton>
                          </Toolbar>
                          <Divider />
                          <Box py={2} px={3}>
                            <Grid container spacing={2}>
                              <Grid item xs={12} sm={6} md={4}>
                                <Typography variant="body2" color="textSecondary">
                                  Agency Project ID
                                </Typography>
                                <Typography variant="body1">{fundingSource.agency_project_id}</Typography>
                              </Grid>
                              <Grid item xs={12} sm={6} md={4}>
                                <Typography variant="body2" color="textSecondary">
                                  Funding Amount
                                </Typography>
                                <Typography variant="body1">
                                  {getFormattedAmount(fundingSource.funding_amount)}
                                </Typography>
                              </Grid>
                              <Grid item xs={12} sm={6} md={4}>
                                <Typography variant="body2" color="textSecondary">
                                  Start / End Date
                                </Typography>
                                <Typography variant="body1">
                                  {getFormattedDateRangeString(
                                    DATE_FORMAT.ShortMediumDateFormat,
                                    fundingSource.start_date,
                                    fundingSource.end_date
                                  )}
                                </Typography>
                              </Grid>
                            </Grid>
                          </Box>
                        </Paper>
                      </ListItem>
                    );
                  })}
                </List>
              </Box>
            )}
          />
        </Box>
      </Box>
    </form>
  );
};

export default ProjectFundingForm;

export const getCodeValueNameByID = (codeSet: IMultiAutocompleteFieldOption[], codeValueId: number): string => {
  if (!codeSet?.length || !codeValueId) {
    return '';
  }

  return codeSet.find((item) => item.value === codeValueId)?.label || '';
};
