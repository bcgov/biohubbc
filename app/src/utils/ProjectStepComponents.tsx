import ProjectCoordinatorForm from 'features/projects/components/ProjectCoordinatorForm';
import ProjectDetailsForm from 'features/projects/components/ProjectDetailsForm';
import ProjectFundingForm from 'features/projects/components/ProjectFundingForm';
import ProjectIUCNForm from 'features/projects/components/ProjectIUCNForm';
import ProjectLocationForm from 'features/projects/components/ProjectLocationForm';
import ProjectObjectivesForm from 'features/projects/components/ProjectObjectivesForm';
import ProjectPartnershipsForm from 'features/projects/components/ProjectPartnershipsForm';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';
import React from 'react';

export interface IProjectStepComponentsProps {
  component: string;
  codes: IGetAllCodeSetsResponse;
}

/**
 * Determine the appropriate component to return for the project step
 *
 * @return {*}
 */
const ProjectStepComponents: React.FC<IProjectStepComponentsProps> = (props) => {
  const { component, codes } = props;

  return (
    <>
      {component === 'ProjectCoordinator' && (
        <ProjectCoordinatorForm
          coordinator_agency={
            codes?.coordinator_agency?.map((item) => {
              return item.name;
            }) || []
          }
        />
      )}

      {component === 'ProjectDetails' && (
        <ProjectDetailsForm
          project_type={
            codes?.project_type?.map((item) => {
              return { value: item.id, label: item.name };
            }) || []
          }
          activity={
            codes?.activity?.map((item) => {
              return { value: item.id, label: item.name };
            }) || []
          }
        />
      )}

      {component === 'ProjectObjectives' && <ProjectObjectivesForm />}

      {component === 'ProjectLocation' && <ProjectLocationForm />}

      {component === 'ProjectIUCN' && (
        <ProjectIUCNForm
          classifications={
            codes?.iucn_conservation_action_level_1_classification?.map((item) => {
              return { value: item.id, label: item.name };
            }) || []
          }
          subClassifications1={
            codes?.iucn_conservation_action_level_2_subclassification?.map((item) => {
              return { value: item.id, iucn1_id: item.iucn1_id, label: item.name };
            }) || []
          }
          subClassifications2={
            codes?.iucn_conservation_action_level_3_subclassification?.map((item) => {
              return { value: item.id, iucn2_id: item.iucn2_id, label: item.name };
            }) || []
          }
        />
      )}

      {component === 'ProjectFunding' && (
        <ProjectFundingForm
          funding_sources={
            codes?.funding_source?.map((item) => {
              return { value: item.id, label: item.name };
            }) || []
          }
          investment_action_category={
            codes?.investment_action_category?.map((item) => {
              return { value: item.id, fs_id: item.fs_id, label: item.name };
            }) || []
          }
        />
      )}

      {component === 'ProjectPartnerships' && (
        <ProjectPartnershipsForm
          first_nations={
            codes?.first_nations?.map((item) => {
              return { value: item.id, label: item.name };
            }) || []
          }
          stakeholder_partnerships={
            codes?.funding_source?.map((item) => {
              return { value: item.name, label: item.name };
            }) || []
          }
        />
      )}
    </>
  );
};

export default ProjectStepComponents;
