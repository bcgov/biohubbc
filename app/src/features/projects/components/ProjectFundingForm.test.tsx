import { render, fireEvent, act, waitFor } from '@testing-library/react';
import { IMultiAutocompleteFieldOption } from 'components/fields/MultiAutocompleteFieldVariableSize';
import { Formik } from 'formik';
import React from 'react';
import ProjectFundingForm, {
  IInvestmentActionCategoryOption,
  IProjectFundingForm,
  ProjectFundingFormInitialValues,
  ProjectFundingFormYupSchema
} from './ProjectFundingForm';
import { codes } from 'test-helpers/code-helpers';
import ProjectStepComponents from 'utils/ProjectStepComponents';

const funding_sources: IMultiAutocompleteFieldOption[] = [
  {
    value: 1,
    label: 'agency 1'
  },
  {
    value: 2,
    label: 'agency 2'
  },
  {
    value: 3,
    label: 'agency 3'
  }
];

const investment_action_category: IInvestmentActionCategoryOption[] = [
  {
    value: 1,
    fs_id: 1,
    label: 'action 1'
  },
  {
    value: 2,
    fs_id: 2,
    label: 'category 1'
  },
  {
    value: 3,
    fs_id: 3,
    label: 'not applicable'
  }
];

describe('ProjectFundingForm', () => {
  it('renders correctly with default empty values', () => {
    const { baseElement } = render(
      <Formik
        initialValues={ProjectFundingFormInitialValues}
        validationSchema={ProjectFundingFormYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async () => {}}>
        {() => (
          <ProjectFundingForm
            funding_sources={funding_sources}
            investment_action_category={investment_action_category}
          />
        )}
      </Formik>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('renders correctly with existing funding values', () => {
    const existingFormValues: IProjectFundingForm = {
      funding_sources: [
        {
          id: 11,
          agency_id: 1,
          investment_action_category: 1,
          investment_action_category_name: 'Action 23',
          agency_project_id: '111',
          funding_amount: 222,
          start_date: '2021-03-14',
          end_date: '2021-04-14',
          revision_count: 23
        }
      ]
    };

    const { baseElement } = render(
      <Formik
        initialValues={existingFormValues}
        validationSchema={ProjectFundingFormYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async () => {}}>
        {() => <ProjectStepComponents component="ProjectFunding" codes={codes} />}
      </Formik>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('shows add funding source dialog on add click', async () => {
    const existingFormValues: IProjectFundingForm = {
      funding_sources: [
        {
          id: 11,
          agency_id: 1,
          investment_action_category: 1,
          investment_action_category_name: 'action 1',
          agency_project_id: '111',
          funding_amount: 222,
          start_date: '2021-03-14',
          end_date: '2021-04-14',
          revision_count: 23
        },
        {
          id: 12,
          agency_id: 2,
          investment_action_category: 2,
          investment_action_category_name: 'category 1',
          agency_project_id: '112',
          funding_amount: 223,
          start_date: '2021-03-15',
          end_date: '2021-04-15',
          revision_count: 24
        }
      ]
    };

    const { getByTestId, queryByText } = render(
      <Formik
        initialValues={existingFormValues}
        validationSchema={ProjectFundingFormYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async () => {}}>
        {() => <ProjectStepComponents component="ProjectFunding" codes={codes} />}
      </Formik>
    );

    const addButton = getByTestId('add-button');

    expect(addButton).toBeInTheDocument();

    fireEvent.click(addButton);

    await waitFor(() => {
      expect(queryByText('Agency Details')).toBeInTheDocument();
    });
  });

  it('shows edit funding source dialog on edit click', async () => {
    await act(async () => {
      const existingFormValues: IProjectFundingForm = {
        funding_sources: [
          {
            id: 11,
            agency_id: 1,
            investment_action_category: 1,
            investment_action_category_name: 'action 1',
            agency_project_id: '111',
            funding_amount: 222,
            start_date: '2021-03-14',
            end_date: '2021-04-14',
            revision_count: 23
          }
        ]
      };

      const { getByTestId, getByText, queryByText } = render(
        <Formik
          initialValues={existingFormValues}
          validationSchema={ProjectFundingFormYupSchema}
          validateOnBlur={true}
          validateOnChange={false}
          onSubmit={async () => {}}>
          {() => <ProjectStepComponents component="ProjectFunding" codes={codes} />}
        </Formik>
      );

      const editButton = await getByTestId('edit-button-0');
      expect(editButton).toBeInTheDocument();

      fireEvent.click(editButton);

      expect(await queryByText('Agency Details')).toBeInTheDocument();

      const cancelButton = await getByText('Cancel');
      expect(cancelButton).toBeInTheDocument();
      fireEvent.click(cancelButton);
      expect(await queryByText('Cancel')).not.toBeInTheDocument();

      fireEvent.click(editButton);
      const saveButton = await getByText('Save Changes');
      expect(saveButton).toBeInTheDocument();
      fireEvent.click(saveButton);
    });
  });

  it('deletes funding source dialog on delete click', async () => {
    await act(async () => {
      const existingFormValues: IProjectFundingForm = {
        funding_sources: [
          {
            id: 11,
            agency_id: 1,
            investment_action_category: 1,
            investment_action_category_name: 'action 1',
            agency_project_id: '111',
            funding_amount: 222,
            start_date: '2021-03-14',
            end_date: '2021-04-14',
            revision_count: 23
          }
        ]
      };

      const { getByTestId, queryByTestId } = render(
        <Formik
          initialValues={existingFormValues}
          validationSchema={ProjectFundingFormYupSchema}
          validateOnBlur={true}
          validateOnChange={false}
          onSubmit={async () => {}}>
          {() => <ProjectStepComponents component="ProjectFunding" codes={codes} />}
        </Formik>
      );

      const deleteButton = await getByTestId('delete-button-0');
      expect(deleteButton).toBeInTheDocument();
      fireEvent.click(deleteButton);

      expect(await queryByTestId('delete-button-0')).not.toBeInTheDocument();
    });
  });
});
