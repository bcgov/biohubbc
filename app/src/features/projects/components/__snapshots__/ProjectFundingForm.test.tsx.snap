// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectFundingForm renders correctly with default empty values 1`] = `
<body>
  <div>
    <form>
      <div
        class="MuiBox-root MuiBox-root-6"
      >
        <header
          class="MuiBox-root MuiBox-root-7"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3"
          >
            Funding Sources (
            0
            )
          </h3>
          <button
            aria-label="Add Funding Source"
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            data-testid="add-button"
            tabindex="0"
            title="Add Funding Source"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  role="presentation"
                  style="width: 1.5rem; height: 1.5rem;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
                    style="fill: currentColor;"
                  />
                </svg>
              </span>
              Add Funding Source
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </header>
        <div
          class="MuiBox-root MuiBox-root-8"
        >
          <div
            class="MuiBox-root MuiBox-root-9"
          >
            <ul
              class="MuiList-root MuiList-dense"
            >
              <div
                class="MuiPaper-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiBox-root MuiBox-root-10"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2"
                  >
                    No Funding Sources
                  </h6>
                </div>
              </div>
            </ul>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;

exports[`ProjectFundingForm renders correctly with existing funding values 1`] = `
<body>
  <div>
    <form>
      <div
        class="MuiBox-root MuiBox-root-16"
      >
        <header
          class="MuiBox-root MuiBox-root-17"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3"
          >
            Funding Sources (
            1
            )
          </h3>
          <button
            aria-label="Add Funding Source"
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            data-testid="add-button"
            tabindex="0"
            title="Add Funding Source"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  role="presentation"
                  style="width: 1.5rem; height: 1.5rem;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
                    style="fill: currentColor;"
                  />
                </svg>
              </span>
              Add Funding Source
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </header>
        <div
          class="MuiBox-root MuiBox-root-18"
        >
          <div
            class="MuiBox-root MuiBox-root-19"
          >
            <ul
              class="MuiList-root MuiList-dense"
            >
              <li
                class="MuiListItem-root makeStyles-fundingListIem-13 MuiListItem-dense MuiListItem-gutters"
              >
                <div
                  class="MuiPaper-root makeStyles-fundingListItemInner-14 MuiPaper-elevation1 MuiPaper-rounded"
                >
                  <div
                    class="MuiToolbar-root MuiToolbar-regular makeStyles-fundingListItemToolbar-15 MuiToolbar-gutters"
                  >
                    <p
                      class="MuiTypography-root makeStyles-title-11 MuiTypography-body1"
                    >
                      Funding source code
                      <span
                        class="makeStyles-titleDesc-12"
                      >
                        (
                        Investment action category
                        )
                      </span>
                    </p>
                    <button
                      aria-label="Edit Funding Source"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                      data-testid="edit-button-0"
                      tabindex="0"
                      title="Edit Funding Source"
                      type="button"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          role="presentation"
                          style="width: 1.5rem; height: 1.5rem;"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                            style="fill: currentColor;"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                    <button
                      aria-label="Remove Funding Source"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                      data-testid="delete-button-0"
                      tabindex="0"
                      title="Remove Funding Source"
                      type="button"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          role="presentation"
                          style="width: 1.5rem; height: 1.5rem;"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z"
                            style="fill: currentColor;"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <hr
                    class="MuiDivider-root"
                  />
                  <div
                    class="MuiBox-root MuiBox-root-20"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                        >
                          Agency Project ID
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1"
                        >
                          111
                        </p>
                      </div>
                      <div
                        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                        >
                          Funding Amount
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1"
                        >
                          $222
                        </p>
                      </div>
                      <div
                        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                        >
                          Start / End Date
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1"
                        >
                          Mar 14, 2021 - Apr 14, 2021
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;
